### PR TITLE
also define $FC and $FCFLAGS in build environment

### PR DESCRIPTION
**Pull request by boegel**
fix for #1392

### CODE REVIEW

**Comment by boegel** [comment]
@geimer: can you take a look at this before I trim the edges?

Should we also define `$F77FLAGS`, while we're at it?

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by geimer** [comment]
@boegel: I can take a look, however, this definitely won't happen before next week. And even this may be wishful thinking...

**Comment by geimer** [comment]
@boegel Regarding `$F77FLAGS`: GNU autotools will use `$FFLAGS` for `$F77` and `$FCFLAGS` for `$FC`.  The `$F90` variable and the associated `$F90FLAGS` are somewhat non-standard (if one considers autotools to be standard...) and custom project-specific extensions. I've never seen `$F77FLAGS` anywhere, though that doesn't mean that it isn't used.

File: easybuild/tools/toolchain/compiler.py [hunk 1]
```diff
@@ -237,21 +238,15 @@ def _set_compiler_flags(self):
...
-
-        self.variables.nappend('F90FLAGS', flags)
-        self.variables.nappend('F90FLAGS', fflags)
-        self.variables.join('F90FLAGS', 'OPTFLAGS', 'PRECFLAGS')
+        for var in ['CFLAGS', 'CXXFLAGS']:
+            self.variables.nappend(var, flags)
+            self.variables.nappend(var, cflags)
+            self.variables.join(var, 'OPTFLAGS', 'PRECFLAGS')
+
+        for var in ['FFLAGS', 'F90FLAGS']:  # FIXME F77FLAGS?
```
Comment (boegel) [inline easybuild/tools/toolchain/compiler.py]: @geimer: so, we should also define `$FCFLAGS` and `$F77FLAGS`?

**Comment by geimer** [comment]
@boegel: `$FCFLAGS` for sure.  As I said, I personally haven't seen `$F77FLAGS` in the wild, but others may know more. At this point I would be reluctant to also add `$F77FLAGS`.

**Comment by boegel** [comment]
OK, I guess we'll stick to just `$FCFLAGS` then, thanks for your input @geimer! I'll try and finish this up in time for inclusion in EB v2.4.0

File: easybuild/toolchains/mpi/mpich.py [hunk 2]
```diff
@@ -46,19 +47,18 @@ class Mpich(Mpi):
 
     MPI_LIBRARY_NAME = 'mpich'
 
+    # FIXME version-dependent? see http://www.mpich.org/static/docs/v3.1.4/www1/mpifort.html
+    MPI_COMPILER_MPIFC = 'mpif90'
+    #MPI_COMPILER_MPIFC = 'mpifort'
```
Comment (boegel) [inline easybuild/toolchains/mpi/mpich.py]: @geimer: it seems like MPICH switched from `mpif90` to `mpifort` in v3.1.2? Any idea about MPICH2?

Comment (geimer) [inline easybuild/toolchains/mpi/mpich.py]: That's right. They suddenly switch the name in a **patch release**! This also makes the sanity check fail when using a v3.1.1 easyconfig with `--try-software-version`. With respect to MPICH2, there is no need to worry as the entire release series provided `mpif90`.

Note that the `mpifort` thing also applies to recent versions of Open MPI (see https://www.open-mpi.org/faq/?category=mpi-apps#mpifort-vs-mpif77-and-mpif90), though they automatically create symlinks to `mpif77` and `mpif90` for backwards compatibility.

**Comment by boegel** [comment]
Summary of the changes made in here (since quite a bit of cleanup was done too):

* also define `$FC` (same as `$F90`) in build environment (same for `$MPIFC`)
* also define `$FCFLAGS` (same as `$F90FLAGS)` in build environment
* define `$MPICF*` for MPICH depending on MPICH version
  * `mpif77`/`mpifort` for MPICH v3.1.2 and newer
  * `mpif77`/`mpif90` for older MPCIH versions
* define `$MPICF*` for OpenMPI depending on OpenMPI version
  * `mpifort` for OpenMPI v1.7 and newer
  * `mpif77`/`mpif90` for older OpenMPI versions

@geimer: how does this sound?

File: easybuild/toolchains/mpi/craympich.py [hunk 3]
```diff
@@ -54,6 +55,7 @@ class CrayMPICH(Mpi):
         '_opt_MPICXX': '',
         '_opt_MPIF77': '',
         '_opt_MPIF90': '',
+        '_opt_MPIFC': '',
```
Comment (geimer) [inline easybuild/toolchains/mpi/craympich.py]: See `easybuild/toolchains/mpi/mpich.py` below: There you use a dict and `MPI_COMPILER_VARIABLES`. Consistency would be nice and using a dict is probably less error prone.

Comment (geimer) [inline easybuild/toolchains/mpi/craympich.py]: Forget it.  In `mpich.py` it's defined at runtime...

File: easybuild/tools/toolchain/constants.py [hunk 4]
```diff
@@ -39,11 +39,13 @@
...
     ('F77', 'Fortran 77 compiler'),
     ('F90', 'Fortran 90 compiler'),
+    ('FC', 'Fortran compiler'),
 ]
 
 COMPILER_FLAGS = [
     ('CFLAGS', 'C compiler flags'),
     ('CXXFLAGS', 'C++ compiler flags'),
+    ('FCFLAGS', 'Fortran compiler flags'),
     ('FFLAGS', 'Fortran compiler flags'),
```
Comment (geimer) [inline easybuild/tools/toolchain/constants.py]: FFLAGS = Fortran 77 compiler flags?

File: easybuild/tools/toolchain/constants.py [hunk 5]
```diff
@@ -72,10 +74,12 @@
         ('CUDA_CXX', 'CUDA C++ compiler command'),
         ('CUDA_F77', 'CUDA Fortran77 compiler command'),
         ('CUDA_F90', 'CUDA Fortran90 compiler command'),
```
Comment (geimer) [inline easybuild/tools/toolchain/constants.py]: CUDA Fortran 77/90 compiler command (space after 'Fortran')?

File: easybuild/tools/toolchain/constants.py [hunk 6]
```diff
@@ -72,10 +74,12 @@
...
+        ('CUDA_FC', 'CUDA Fortran compiler command'),
     ],
     FlagList: [
         ('CUDA_CFLAGS', 'CUDA C compiler flags'),
         ('CUDA_CXXFLAGS', 'CUDA C++ compiler flags'),
+        ('CUDA_FCFLAGS', 'CUDA Fortran compiler flags'),
         ('CUDA_FFLAGS', 'CUDA Fortran compiler flags'),
```
Comment (geimer) [inline easybuild/tools/toolchain/constants.py]: CUDA Fortran 77 compiler flags?

File: test/framework/toolchain.py [hunk 7]
```diff
@@ -189,7 +169,7 @@ def test_validate_pass_by_value(self):
     def test_optimization_flags(self):
         """Test whether optimization flags are being set correctly."""
 
-        flag_vars = ['CFLAGS', 'CXXFLAGS', 'FFLAGS', 'F90FLAGS']
+        flag_vars = ['CFLAGS', 'CXXFLAGS', 'FCFLAGS', 'FFLAGS', 'F90FLAGS']
```
Comment (geimer) [inline test/framework/toolchain.py]: It's only a test so it doesn't really matter, but at basically all other places you used the order F77, F90, FC. Maybe adopt it here (and 6 more times below) as well?

**Comment by geimer** [comment]
@boegel: Visually looks good to me; my comments are really nitpicking ;-)  I've also tried to build Open MPI 1.10.0 using the Intel compilers on top of this PR w/o having FC set in `configopts` (like in https://github.com/hpcugent/easybuild-framework/issues/1392) and the build worked out fine.

File: easybuild/toolchains/mpi/craympich.py (hunk 3 above)
Comment (boegel) [inline easybuild/toolchains/mpi/craympich.py]: No, I think this is a valid remark.

File: test/framework/toolchain.py (hunk 7 above)
Comment (boegel) [inline test/framework/toolchain.py]: Like you said, it doesn't really matter. In fact, the order here matches the order of the `COMPILER_FLAGS` list.

**Comment by boegel** [comment]
Going in, thanks for the review @geimer!

**Pull request merged**